providers/base/fetch.go

```diff
@@ -15,57 +15,44 @@ import (
 	providertypes "github.com/skip-mev/slinky/providers/types"
 )
 
-// fetch is the main blocker for the provider. It is responsible for fetching data from the
-// data provider and updating the data.
+// fetch is the main blocker for the provider. It is responsible for fetching data from
+// the data provider and updating the data. Note that the context passed here is valid
+// until either the parent context (provider's main context) is cancelled, the fetch routine
+// encounters an error, or the provider is manually stopped.
 func (p *Provider[K, V]) fetch(ctx context.Context) error {
-	// responseCh is used to receive the response(s) from the query handler.
-	var responseCh chan providertypes.GetResponse[K, V]
-	switch {
-	case p.api != nil:
-		// If the provider is an API provider, then the buffer size is set to the number of IDs.
-		responseCh = make(chan providertypes.GetResponse[K, V], len(p.GetIDs()))
-	case p.ws != nil:
-		// Otherwise, the buffer size is set to the max buffer size configured for the websocket.
-		responseCh = make(chan providertypes.GetResponse[K, V], p.wsCfg.MaxBufferSize)
-	default:
-		return fmt.Errorf("no api or websocket configured")
-	}
-
-	// Start the receive loop.
-	go p.recv(ctx, responseCh)
-
 	// Determine which loop to use based on whether the provider is an API or webSocket provider.
 	switch {
-	case p.api != nil:
-		return p.startAPI(ctx, responseCh)
-	case p.ws != nil:
-		return p.startMultiplexWebsocket(ctx, responseCh)
+	case p.Type() == providertypes.API:
+		return p.startAPI(ctx)
+	case p.Type() == providertypes.WebSockets:
+		return p.startMultiplexWebsocket(ctx)
 	default:
 		return fmt.Errorf("no api or websocket configured")
 	}
 }
 
 // startAPI is the main loop for the provider. It is responsible for fetching data from the API
 // and updating the data.
-func (p *Provider[K, V]) startAPI(ctx context.Context, responseCh chan<- providertypes.GetResponse[K, V]) error {
+func (p *Provider[K, V]) startAPI(ctx context.Context) error {
 	p.logger.Info("starting api query handler")
 
 	// Start the data update loop.
+	handler := p.GetAPIHandler()
+	ids := p.GetIDs()
```

Comment:
It seems like the problem you're running into is keeping state in sync between the `IDs` in the base provider and the `ProviderMarketmap` in the data handler.

My sense is that all of this would be much easier if you just moved all state (i.e. the `ProviderMarketMap`) to the base provider and made any `APIDataHandler` functionality stateless. 

I think that would completely eliminate the problem of having to keep state in sync between fetch and main provider.

Overall though using contexts to manage stop/start is much cleaner than the message handler channel based stuff we had before.